### PR TITLE
feat: use ts to parse tsconfig

### PR DESCRIPTION
**Pull request by dominikg**
eval gives me pause and doesn't actually resolve tsconfig files that extend others or resolve paths or return values matching ts.CompilerOptions.

one test fails, it seems to be due to the fact that module is  enum 99 instead of 'esnext', though i don't understand why that happens, the enum is actually more correct according to ts.CompilerOptions

Note that `parse_tsconfig` returns `"declarationDir": "test/types"`  for `test/tsconfig.json` where eval returned just `types`. Doesn't matter in this case as declarationDir is reset to undefined, but for other paths it might be relevant.

### CODE REVIEW

**Comment by dominikg** [comment]
running this branch on sveltekit shows a lot of external types going missing too, so it's not just this testcase here.

File: src/utils.js [hunk 1]
```diff
@@ -434,3 +434,34 @@ export function is_reference(node) {
 
 	return true;
 }
+
+/**
+ * parse tsconfig.json with typescript api
+ *
```
Comment (benmccann) [inline src/utils.js:440]: I think we've generally not put a blank line between description and params
```suggestion
```

File: src/utils.js [hunk 2]
```diff
@@ -434,3 +434,34 @@ export function is_reference(node) {
...
+ * @param {string} tsconfig_file
+ * @returns {{include: string[]|undefined, exclude: string[]|undefined,compilerOptions: ts.CompilerOptions}}}
+ * @throws {Error} if ts api returns error diagnostics
+ */
+export function parse_tsconfig(tsconfig_file) {
+	const {config,error: read_diagnostic} = ts.readConfigFile(tsconfig_file, ts.sys.readFile)
+	/**@type {(error?: ts.Diagnostic)=> boolean} */
```
Comment (benmccann) [inline src/utils.js:447]:
```suggestion
	/** @type {(error?: ts.Diagnostic)=> boolean} */
```

File: src/utils.js [hunk 3]
```diff
@@ -434,3 +434,34 @@ export function is_reference(node) {
...
+	const isError = (d) => d?.category === ts.DiagnosticCategory.Error;
+	if(isError(read_diagnostic)) {
+		throw new Error(`failed to read ${tsconfig_file}. ${read_diagnostic?.messageText}`)
+	}
+	const {raw,options,errors: parse_diagnostics} = ts.parseJsonConfigFileContent(
```
Comment (benmccann) [inline src/utils.js:452]: You were right about needing prettier setup in this repo 😆 
```suggestion
	const { raw, options, errors: parse_diagnostics } = ts.parseJsonConfigFileContent(
```

File: src/utils.js [hunk 4]
```diff
@@ -434,3 +434,34 @@ export function is_reference(node) {
...
+		config,
+		ts.sys,
+		path.dirname(tsconfig_file)
+	)
+	const parse_errors = parse_diagnostics?.filter(isError)
+	if(parse_errors.length > 0) {
```
Comment (benmccann) [inline src/utils.js:458]:
```suggestion
	if(parse_errors.length > 0) {
```

Comment (benmccann) [inline src/utils.js:458]:
```suggestion
	if(parse_errors.length > 0) {
```

File: src/utils.js [hunk 5]
```diff
@@ -434,3 +434,34 @@ export function is_reference(node) {
 
 	return true;
 }
+
+/**
+ * parse tsconfig.json with typescript api
+ *
+ * @param {string} tsconfig_file
+ * @returns {{include: string[]|undefined, exclude: string[]|undefined,compilerOptions: ts.CompilerOptions}}}
```
Comment (benmccann) [inline src/utils.js:442]: It looks like you've got two opening braces, but three closing braces. Also, Prettier would put each property on its own line here
```suggestion
 * @returns {{
 *   include: string[]|undefined,
 *   exclude: string[]|undefined,
 *   compilerOptions: ts.CompilerOptions
 * }}
```

File: src/index.js [hunk 6]
```diff
@@ -4,7 +4,7 @@ import ts from 'typescript';
 import MagicString from 'magic-string';
 import { getLocator } from 'locate-character';
 import { SourceMapGenerator } from '@jridgewell/source-map';
-import { clean_jsdoc, get_input_files, is_declaration, resolve_dts, walk, write } from './utils.js';
+import {clean_jsdoc, get_input_files, is_declaration, parse_tsconfig, resolve_dts, walk, write} from './utils.js';
```
Comment (benmccann) [inline src/index.js:7]:
```suggestion
import { clean_jsdoc, get_input_files, is_declaration, parse_tsconfig, resolve_dts, walk, write } from './utils.js';
```

**Comment by Rich-Harris** [comment]
the reason this is failing is that the resulting `compilerOptions` has a `module: 99` and a `target: 99` instead of `"esnext"`. i have no idea why that would affect anything.

**Comment by dominikg** [comment]
yeah, i have no clue. in tsconfck i'm doing a bunch of things to get the output of typescripts own parser to match what the content of an equivalent tsconfig.json file would look like https://github.com/dominikg/tsconfck/blob/main/src/parse-native.ts#L163

so we could just add similar reverse replace 99 with esnext here and call it a day :man_cartwheeling:

**Comment by dominikg** [comment]
> yeah, i have no clue. in tsconfck i'm doing a bunch of things to get the output of typescripts own parser to match what the content of an equivalent tsconfig.json file would look like https://github.com/dominikg/tsconfck/blob/main/src/parse-native.ts#L163
> 
> so we could just add similar reverse replace 99 with esnext here and call it a day 🤸‍♂️

@Rich-Harris 
turns out it also starts working if you don't override moduleResolution with undefined.  What was the rationale behind that?

**Comment by Rich-Harris** [comment]
> What was the rationale behind that?

I don't recall, honestly. Let's try removing it and seeing what happens...

**Pull request merged**